### ISSUE TITLE
meilisearch "dies" when adding documents

### ISSUE THREAD

**Issue by funlennysub** [meilisearch/meilisearch#4903]
**Describe the bug**
I have a CLI app that has an option to sync meilisearch with postgres table. It first deletes every document from the index using `index.delete_all_documents()` and waits for the completion, then it gets all the data from postres and performs `index.add_documents_in_batches(&[..], Some(7_000), Some("id"))` using rust sdk. For me this problem occurred at `1,498,000` documents and somewhere at 2.5mil, trying to cancel all `documentAdditionOrUpdate` tasks doesn't work until you restart meilisearch. Doing what I do for the sync after restart seems to be working, but it will break again if you run it after everything is done.

**To Reproduce**
1. Populate index with around 20mil documents (That's how many I have)
2. Perform [delete all](https://www.meilisearch.com/docs/reference/api/documents#delete-all-documents) and wait for completion
3. Insert 20mil documents with 7k per batch
4. Observe

**Expected behavior**
Every task should finish and insert everything

**Screenshots**
Last completed task:
![image](https://github.com/user-attachments/assets/9df1bd17-7a8f-4595-a90b-415208f5ee50)

Next task:
![image](https://github.com/user-attachments/assets/4646a310-ba5b-4bf1-8279-d2289a958ec2)

**Meilisearch version:**
v1.10.0

**Additional context**
- When it stuck global stats report `"isIndexing": true`
- Meilisearch is running in docker

**Comment by funlennysub** [comment]
Right, my bad, when it works as expected it finishes in 20-30 minutes, but I left it multiple times in a broken state for hours and it still was stuck at same number of documents and no new tasks were finished. 

> As a reminder, Meilisearch is asynchronous, from the screens you posted it looks like Meilisearch is currently indexing documents.
> 
I know, this makes it weirder because according to the documentation, cancellation tasks should have priority yet they only cancel after you restart meilisearch

Also btop reports some cpu activity (10%) but as I said nothing happens.

**Comment by dureuill** [comment]
Apologies, but I really need to understand what is going on here.

> I left it multiple times in a broken state for hours

What do you mean by "in a broken state"?

> and it still was stuck at same number of documents and no new tasks were finished.

Due to [auto-batching](https://www.meilisearch.com/docs/learn/getting_started/documents#auto-batching), if multiple tasks can be processed at the same time, then they will be processed at the same time. This increases throughput (more documents will have been indexed per second when the batch of tasks completes), but decreases latency (an individual task may complete later than if processed without auto batching).

You can check how many tasks are autobatching with `/tasks?statuses=processing`, this `"total"` field will contain this value.

**Comment by funlennysub** [comment]
> What do you mean by "in a broken state"?

By broken state I mean meilisearch not processing tasks.
I'm also pretty sure I would see more activity if it was actually indexing etc in btop but it just sits at 2gb ram and 10% cpu usage for multiple hours.
In a "working state" meilisearch activity would look like 10gb ram and 50% cpu usage and takes appropriate amount of time

**Comment by funlennysub** [comment]
> Can you check how many tasks are autobatching with /tasks?statuses=processing when this happens?

Right after i took the screenshot below it said 280, now it says 3251 (I add ~25mil with 5k in each batch).
And i have the same behavior as i described, 2gb ram and ~10% cpu usage and sometimes disk writes at ~65M/s and reads at 20M/s according to iotop

![image](https://github.com/user-attachments/assets/4c491857-d641-487c-be0c-5a4a668df11a)

**Comment by funlennysub** [comment]
Okay this time it actually inserted everything, I will try again

**Comment by funlennysub** [comment]
I tried like 6 more times, and only got it once when I set my date back to august 31 ([who knows](https://bugs.launchpad.net/ubuntu/+source/file/+bug/248619)) and it just was going back and forth between 0K/s and 7000K/s disk reads, but when I tried again on aug 31 it worked just fine.

I just find it weird that when I left it run for a night it still was doing something when I woke up. I will keep monitoring it and if anything happens, I will mention it here. 

Feel free to close the issue
